flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManager.java

```diff
@@ -0,0 +1,340 @@
+			defaultMemoryMB,
+			defaultCpus,
+			numPendingPodRequests);
+
+		final String podName = String.format(
+			TASK_MANAGER_POD_FORMAT,
+			clusterId,
+			currentMaxAttemptId.get(),
+			currentMaxPodId.addAndGet(1));
+
+		final HashMap<String, String> env = new HashMap<>();
+		env.put(Constants.ENV_FLINK_POD_NAME, podName);
+		env.putAll(taskManagerParameters.taskManagerEnv());
+
+		final TaskManagerPodParameter parameter = new TaskManagerPodParameter(
+			podName,
+			taskManagerStartCommand,
+			defaultMemoryMB,
+			defaultCpus,
+			env);
+
+		log.info("TaskManager {} will be started with {}.", podName, taskExecutorResourceSpec);
+		kubeClient.createTaskManagerPod(parameter);
+	}
+
+	/**
+	 * Request new pod if pending pods cannot satisfy pending slot requests.
+	 */
+	private void requestKubernetesPodIfRequired() {
+		final int requiredTaskManagerSlots = getNumberRequiredTaskManagerSlots();
+		final int pendingTaskManagerSlots = numPendingPodRequests * numSlotsPerTaskManager;
+
+		if (requiredTaskManagerSlots > pendingTaskManagerSlots) {
+			requestKubernetesPod();
+		}
+	}
+
+	private void removePodIfTerminated(KubernetesPod pod) {
+		if (pod.isTerminated()) {
+			kubeClient.stopPod(pod.getName());
+			final KubernetesWorkerNode kubernetesWorkerNode = workerNodes.remove(new ResourceID(pod.getName()));
+			if (kubernetesWorkerNode != null) {
+				requestKubernetesPodIfRequired();
+			}
+		}
+	}
+
+	private List<String> getTaskManagerStartCommand() {
+
+		final String confDir = flinkConfig.getString(KubernetesConfigOptions.FLINK_CONF_DIR);
+		final boolean hasLogback = new File(confDir, Constants.CONFIG_FILE_LOGBACK_NAME).exists();
+		final boolean hasLog4j = new File(confDir, Constants.CONFIG_FILE_LOG4J_NAME).exists();
+
+		final String logDir = flinkConfig.getString(KubernetesConfigOptions.FLINK_LOG_DIR);
+
+		final String mainClassArgs = "--" + CommandLineOptions.CONFIG_DIR_OPTION.getLongOpt() + " " +
+			flinkConfig.getString(KubernetesConfigOptions.FLINK_CONF_DIR) + " " +
+			BootstrapTools.getDynamicProperties(flinkClientConfig, flinkConfig);
+
+		final String command = KubernetesUtils.getTaskManagerStartCommand(
+			flinkConfig,
+			taskManagerParameters,
+			confDir,
+			logDir,
+			hasLogback,
+			hasLog4j,
+			KubernetesTaskExecutorRunner.class.getCanonicalName(),
+			mainClassArgs);
+
+		return Arrays.asList("/bin/bash", "-c", command);
+	}
+
+	/**
+	 * Get task manager labels for the current Flink cluster. They could be used to watch the pods status.
+	 *
+	 * @return Task manager labels.
+	 */
+	private Map<String, String> getTaskManagerLabels() {
+		final Map<String, String> labels = new HashMap<>();
+		labels.put(Constants.LABEL_TYPE_KEY, Constants.LABEL_TYPE_NATIVE_TYPE);
+		labels.put(Constants.LABEL_APP_KEY, clusterId);
+		labels.put(Constants.LABEL_COMPONENT_KEY, Constants.LABEL_COMPONENT_TASK_MANAGER);
+		return labels;
+	}
+
+	protected FlinkKubeClient createFlinkKubeClient() {
```

Comment:
Maybe not need this separate method with only one sentence. If you prefer to maintaining, it should be private method.

Reply:
I will remove the `createFlinkKubeClient()` and add a `setFlinkKubeClient` for test purpose. We need to set the testing kube client to `KubernetesResourceManager` so that it could be used to create k8s resources.